packages/token-auth-middleware/src/lib/handler.ts

```diff
@@ -0,0 +1,282 @@
+import loginPage from '../assets/login.html?raw';
+import notificationPage from '../assets/notification.html?raw';
+import pageStyles from '../assets/styles.css?raw';
+
+function injectStyles(html: string, styles: string) {
+  return html.replace(
+    '<link rel="stylesheet" href="styles.css" />',
+    `<style>${styles}</style>`,
+  );
+}
+
+export type TokenAuthHandlerOptions = {
+  /**
+   * Custom login page to redirect to.
+   */
+  customLogin?: string;
+
+  /**
+   * Lifetime of the login token in seconds.
+   * Defaults to one minute.
+   */
+  tokenLifetime?: number;
+
+  /**
+   * Lifetime of the session in seconds.
+   * Defaults to no expiration.
+   */
+  sessionLifetime?: number;
+};
+
+export class TokenExpiredError extends Error {
+  constructor() {
+    super('Token expired');
+  }
+}
+
+export class TokenInvalidError extends Error {
+  constructor() {
+    super('Token invalid');
+  }
+}
+
+export type Info = Record<string, any>;
+
+export interface TokenEncoderInterface {
+  create(id: string, lifetime?: number): Promise<string>;
+
+  /**
+   * @throws TokenExpiredError
+   * @throws TokenInvalidError
+   * @param token
+   */
+  validate(token: string): Promise<string>;
+}
+
+export interface AuthenticationBackendInterface<TInfo extends Info> {
+  getInfo(id: string): Promise<TInfo | undefined>;
+  deliver(id: string, link: string): Promise<void>;
+}
+
+export function metaRedirect(message: string, destination: string) {
+  return `<html lang="en"><head><title>${message}</title><meta http-equiv="refresh" content="0;url=${destination}" /></head></html>`;
```

Comment:
The message is not an issue, since its hardcoded within the application. But the destination comes from url or headers.
For login links i could encode destination into the JWT payload. And for logout, might just redirect to the login page instead.

Reply:
I think it's easier to escape them once rather than checking if the values are safe every time an eye catches an unescaped html template 🤷